core/WASM_core.md

```diff
@@ -0,0 +1,352 @@
+mentioned at the top of the doc. The only sys interface things it appears we'll be able to reliably
+use are clocks (which is good, that does come up) and possibly file I/O, which currently does not
+happen in Core (and probably never should) except for in the test server usecase.
+
+
+### Core API changes
+Operating under the assumption that all the major syscalls will need to be handled through callbacks
+into lang. In WASM, calling functions in the host is supported via `import`ing those host functions.
+Ultimately these functions may be bundled together into a trait or something on the Core side, but
+this doc will focus on the individual functions.
+
+Here are the functions I think we'll need to add to the Core<-->Lang interface, beyond the existing
+ones that lang already calls. They're defined as they would appear in Rust, after being imported.
+This is definitely a rough draft, as it will be influenced by runtime selection and whatever
+practical problems we run into along the way.
+
+```rust
+// In reality, everything in these functions is going to need to be serialized and shuttled back
+// and forth as bytes. That makes for uninteresting signatures, though, so instead I show what
+// those signatures would look like if we were using deserialized types directly. The exact details
+// of how that serialization happens are intentionally not included here.
+//
+// WASM has an `ExternRef` type which is a pointer to a host-owned opaque object.
+
+// For all the types which wrap an `ExternRef`, the implementations on those types which take
+// `self` will be implemented as calls to a host-defined function which looks like:
+fn func_name(receiver: ExternRef, arg: ArgType) -> ResultType;
+// There may be a more clever way to manage some of this by using `Table` and `Func` types along with
+// `ExternRef` but I need to play around more to determine that.
+
+// -------------- Grpc -----------------
+
+// Would be passed into core along with client config options when creating instantiating a client.
+// Core can then augment the client with retry behavior, and the client can then be used when
+// instantiating Core workers.
+struct GrpcClientRef {
+    /// Points to the host's gRPC client
+    host_ref: ExternRef,
+}
+
+struct GrpcRequest {
+    uri: String,
+    body: Vec<u8>,
+    metadata: Metadata,
+}
+
+struct GrpcResponse {
+    body: Vec<u8>,
+    metadata: Metadata,
+}
+
+struct GrpcStatus {
+    // Serializable equivalent of https://docs.rs/tonic/latest/tonic/struct.Status.html
+}
+
+struct Metadata {
+    data: HashMap<String, Vec<u8>>,
+}
+
+struct GrpcClientConfig {
+    // Details elided. See `ClientConfig` in core. Would *not* include `RetryConfig`, as that will
+    // continue to be handled in core.
+}
+
+impl GrpcClientRef {
+    /// Make a unary grpc call
+    async fn grpc_unary(
+        &self,
+        req: GrpcRequest,
+    ) -> Result<GrpcResponse, GrpcStatus> {}
```

Comment:
I think you'll be simplifying these types when you start making them "extern" to be exported from your WASM module. Specifically, just put the gRPC status into the response.

And of course you'll end up needing a two-function approach so you'll need some kind of request identifier. Langs will impl a "submit gRPC request" and core will impl a "on gRPC response" to be called by lang.